### PR TITLE
Fix fork choice store checkpoints inconsistency issue

### LINKED ISSUE #2515: Fork Choice Store Justified-Finalized Inconsistency
Thank @adiasg and @djrtwo for the discussion.

### Issue
See **[full proof here](https://notes.ethereum.org/@hww/fork-choice-store-inconsistency)**.

In the edge cases, `store.justified_checkpoint` would be reverted by `store.best_justified_checkpoint` and it may be NOT a descendant of `store.finalized_checkpoint`.

Note that this edge case would only happen with more than **1/3rd slashable validators**.

#### Solution 1: Update `store.best_justified_checkpoint` in Case 1 and Case 2 to ensure that  `store.best_justified_checkpoint` is always on the `store.finalized_checkpoint` chain 

That is:
```python
            # Update justified if new justified is later than store justified
            if state.current_justified_checkpoint.epoch > store.justified_checkpoint.epoch:
                # ------------------------
                # >>>>>>>> CASE 1 <<<<<<<<
                # ------------------------
                store.justified_checkpoint = state.current_justified_checkpoint
                store.best_justified_checkpoint = store.justified_checkpoint
                return
```

```python
            # Update justified if store justified is not in chain with finalized checkpoint
            finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
            ancestor_at_finalized_slot = get_ancestor(store, store.justified_checkpoint.root, finalized_slot)
            if ancestor_at_finalized_slot != store.finalized_checkpoint.root:
                # ------------------------
                # >>>>>>>> CASE 2 <<<<<<<<
                # ------------------------
                store.justified_checkpoint = state.current_justified_checkpoint
                store.best_justified_checkpoint = store.justified_checkpoint
                return
```

#### Solution 2: add checks in the `on_tick` function to ensure that `store.justified_checkpoint` will be on the `store.finalized_checkpoint` chain 

```python
def on_tick(store: Store, time: uint64) -> None:
    ...
    # Update store.justified_checkpoint if a better checkpoint is known
    if store.best_justified_checkpoint.epoch > store.justified_checkpoint.epoch:
        # Check if store.best_justified_checkpoint is on the store.finalized_checkpoint chain
        finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)    
        ancestor_at_finalized_slot = get_ancestor(store, store.best_justified_checkpoint.root, finalized_slot)
        if ancestor_at_finalized_slot == store.finalized_checkpoint.root:
            store.justified_checkpoint = store.best_justified_checkpoint
```

### PR DESCRIPTION
**Pull request by hwwhww**
Fix #2515

- Apply solution 2 in #2515 
- Add basic unit tests to cover the new code; the corresponding test suite updates will need the helper functions from #2487. We can go merge any PR first.

### CODE REVIEW

File: tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_tick.py [hunk 1]
```diff
@@ -26,18 +34,92 @@ def test_basic(spec, state):
...
+    # [Mock store.best_justified_checkpoint]
+    # Create a block at epoch 1
+    next_epoch(spec, state)
+    block = build_empty_block_for_next_slot(spec, state)
+    state_transition_and_sign_block(spec, state, block)
+    store.blocks[block.hash_tree_root()] = block.copy()
+    store.block_states[block.hash_tree_root()] = state.copy()
+    parent_block = block.copy()
+    # To make compute_slots_since_epoch_start(current_slot) == 0, transition to the end of the epoch
+    slot = state.slot + spec.SLOTS_PER_EPOCH - (state.slot + spec.SLOTS_PER_EPOCH) % spec.SLOTS_PER_EPOCH - 1
```
Comment (adiasg) [inline tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_tick.py:49]:
```suggestion
    slot = state.slot + spec.SLOTS_PER_EPOCH - state.slot % spec.SLOTS_PER_EPOCH - 1
```

**Review by adiasg — APPROVED** [review]
Minor suggestion, but looks good otherwise!

File: tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_tick.py [hunk 2]
```diff
@@ -26,18 +34,92 @@ def test_basic(spec, state):
...
+    state = init_state.copy()
+    next_epoch(spec, state)
+    block = build_empty_block_for_next_slot(spec, state)
+    block.body.graffiti = b'\x22' * 32
+    state_transition_and_sign_block(spec, state, block)
+    store.blocks[block.hash_tree_root()] = block.copy()
+    store.block_states[block.hash_tree_root()] = state.copy()
+    parent_block = block.copy()
+    # To make compute_slots_since_epoch_start(current_slot) == 0, transition to the end of the epoch
+    slot = state.slot + spec.SLOTS_PER_EPOCH - (state.slot + spec.SLOTS_PER_EPOCH) % spec.SLOTS_PER_EPOCH - 1
```
Comment (adiasg) [inline tests/core/pyspec/eth2spec/test/phase0/unittests/fork_choice/test_on_tick.py:103]:
```suggestion
    slot = state.slot + spec.SLOTS_PER_EPOCH - state.slot % spec.SLOTS_PER_EPOCH - 1
```

**Review by adiasg — APPROVED** [review]

**Review by djrtwo — APPROVED** [review]
I didn't review tests, but the logic in on_tick looks correct to me

**Pull request merged**